swarm/src/protocols_handler/multi.rs

```diff
@@ -0,0 +1,362 @@
+    H: fmt::Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Upgrade")
+            .field("upgrades", &self.upgrades)
+            .finish()
+    }
+}
+
+impl<K, H> UpgradeInfoSend for Upgrade<K, H>
+where
+    K: Hash + Eq + Clone + Send + 'static,
+    H: UpgradeInfoSend
+{
+    type Info = KeyedProtoName<K, H::Info>;
+    type InfoIter = std::vec::IntoIter<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        self.upgrades.iter().map(|(k, i)| iter::repeat(k.clone()).zip(i.protocol_info()))
+            .flatten()
+            .map(|(k, i)| KeyedProtoName(k, i))
+            .collect::<Vec<_>>()
+            .into_iter()
+    }
+}
+
+impl<K, H> InboundUpgradeSend for Upgrade<K, H>
+where
+    H: InboundUpgradeSend,
+    K: Clone + Hash + Eq + Send + 'static
+{
+    type Output = (K, <H as InboundUpgradeSend>::Output);
+    type Error  = (K, <H as InboundUpgradeSend>::Error);
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_inbound(mut self, resource: NegotiatedSubstream, info: Self::Info) -> Self::Future {
+        let KeyedProtoName(key, info) = info;
+        let u = self.upgrades.remove(&key).expect(
+            "`upgrade_inbound` is applied to a key from `protocol_info`, which only contains \
+            keys from the same set of upgrades we are searching here, therefore looking for this \
+            key is guaranteed to give us a non-empty result; qed"
+        );
+        u.upgrade_inbound(resource, info).map(move |out| {
+            match out {
+                Ok(o) => Ok((key, o)),
+                Err(e) => Err((key, e))
+            }
+        })
+        .boxed()
+    }
+}
+
+impl<K, H> OutboundUpgradeSend for Upgrade<K, H>
+where
+    H: OutboundUpgradeSend,
+    K: Clone + Hash + Eq + Send + 'static
+{
+    type Output = (K, <H as OutboundUpgradeSend>::Output);
+    type Error  = (K, <H as OutboundUpgradeSend>::Error);
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(mut self, resource: NegotiatedSubstream, info: Self::Info) -> Self::Future {
+        let KeyedProtoName(key, info) = info;
+        let u = self.upgrades.remove(&key).expect(
+            "`upgrade_outbound` is applied to a key from `protocol_info`, which only contains \
+            keys from the same set of upgrades we are searching here, therefore looking for this \
+            key is guaranteed to give us a non-empty result; qed"
+        );
+        u.upgrade_outbound(resource, info).map(move |out| {
+            match out {
+                Ok(o) => Ok((key, o)),
+                Err(e) => Err((key, e))
+            }
+        })
+        .boxed()
+    }
+}
+
+/// Check that no two protocol names are equal.
+fn uniq_proto_names<I, T>(iter: I) -> Result<(), DuplicateProtoname>
+where
+    I: Iterator<Item = T>,
+    T: UpgradeInfoSend
+{
+    let mut set = std::collections::HashSet::new();
```

Comment:
```suggestion
    let mut set = HashSet::new();
```

Imports on top

Reply:
Maybe you should write a style guide so that contributors can anticipate your requirements.